UndertaleModLib/Models/UndertaleEmbeddedTexture.cs

```diff
@@ -379,152 +349,42 @@ protected void OnPropertyChanged([CallerMemberName] string name = null)
-                {
-                    FormatQOI = true;
-                    FormatBZ2 = true;
-
-                    // Don't really care about the width/height, so skip them, as well as header
-                    reader.Position += (uint)(gm2022_5 ? 12 : 8);
-
-                    // Need to fully decompress and convert the QOI data to PNG for compatibility purposes (at least for now)
-                    if (sharedStream.Length != 0)
-                        sharedStream.Seek(0, SeekOrigin.Begin);
-                    BZip2.Decompress(reader.Stream, sharedStream, false);
-                    ReadOnlySpan<byte> decompressed = sharedStream.GetBuffer().AsSpan()[..(int)sharedStream.Position];
-                    using Bitmap bmp = QoiConverter.GetImageFromSpan(decompressed);
-                    sharedStream.Seek(0, SeekOrigin.Begin);
-                    bmp.Save(sharedStream, ImageFormat.Png);
-                    TextureBlob = new byte[(int)sharedStream.Position];
-                    sharedStream.Seek(0, SeekOrigin.Begin);
-                    sharedStream.Read(TextureBlob, 0, TextureBlob.Length);
-                    return;
-                }
-                else if (header.Take(4).SequenceEqual(QOIHeader))
-                {
-                    FormatQOI = true;
-                    FormatBZ2 = false;
-
-                    // Need to convert the QOI data to PNG for compatibility purposes (at least for now)
-                    using Bitmap bmp = QoiConverter.GetImageFromStream(reader.Stream);
-                    if (sharedStream.Length != 0)
-                        sharedStream.Seek(0, SeekOrigin.Begin);
-                    bmp.Save(sharedStream, ImageFormat.Png);
-                    TextureBlob = new byte[(int)sharedStream.Position];
-                    sharedStream.Seek(0, SeekOrigin.Begin);
-                    sharedStream.Read(TextureBlob, 0, TextureBlob.Length);
-                    return;
-                }
-                else
-                    throw new IOException("Didn't find PNG or QOI+BZip2 header");
-            }
-
-            // There is no length for the PNG anywhere as far as I can see
-            // The only thing we can do is parse the image to find the end
-            while (true)
-            {
-                // PNG is big endian and BinaryRead can't handle that (damn)
-                uint len = (uint)reader.ReadByte() << 24 | (uint)reader.ReadByte() << 16 | (uint)reader.ReadByte() << 8 | (uint)reader.ReadByte();
-                uint type = reader.ReadUInt32();
-                reader.Position += len + 4;
-                if (type == 0x444e4549) // 0x444e4549 -> "IEND"
-                    break;
-            }
-
-            long length = reader.Position - startAddress;
-            reader.Position = startAddress;
-            TextureBlob = reader.ReadBytes((int)length);
+            Image = GMImage.FromBinaryReader(reader, maxEndOfStreamPosition, gm2022_5);
```

Comment:
Not always... only if it's a BZ2+QOI type image, although it could also break on other image types in the future. I guess since it's public I'll add an exception for that, and probably document that behavior.